File: _72_edit_distance.py
"""
72. Edit Distance

Given two words word1 and word2,
find the minimum number of steps required to convert word1 to word2.
(each operation is counted as 1 step.)

You have the following 3 operations permitted on a word:

a) Insert a character
b) Delete a character
c) Replace a character
"""
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        # From word1 -> word2
        if len(word1) == 0:
            return len(word2)

        # minDistance(word1, word2) should be equal to minDistance(word2, word1)
        if len(word1) < len(word2):
            return self.minDistance(word2, word1)

        # Now len(word1) >= len(word2)
        #print "from ", word1, "to ", word2
        prev_dist = range(len(word2) + 1)   # prev_dist[j+1]: distance from word1 to word2[j]
                                            # initially, prev_dist[j] should be j
        for i, c1 in enumerate(word1):
            current_dist = [i + 1]
            for j, c2 in enumerate(word2):
                #print word1[:min(i+1, len(word1))], "from ", word1, " <=> ", word2[:min(j+1, len(word2))], "from ", word2
                insert = prev_dist[j+1] + 1     # j+1 instead of j since prev_dist and current_dist
                delete = current_dist[j] + 1    # are one character longer than word2
                replace = prev_dist[j] + (c1 != c2)
                current_dist.append(min(insert, delete, replace))
                #print "picking...", insert, delete, replace
            prev_dist = current_dist
            #print "current distance: ", current_dist[1:], "(from ", word1[:min(i+1, len(word1))], "to ", word2, ")"
        
        return prev_dist[-1]

File: current_session/c++/72.edit-distance.cpp
/*
 * @lc app=leetcode id=72 lang=cpp
 *
 * [72] Edit Distance
 */

// @lc code=start
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        if (word1.size() == 0 && word2.size() == 0) return 0;
        if (word1.size() == 0) return word2.size();
        if (word2.size() == 0) return word1.size();

        int A = word1.size(), B = word2.size();
        vector<vector<int>> dp(A+1, vector<int>(B+1, 0));
        for (int i = 0; i <= A; ++i) {
        for (int j = 0; j <= B; ++j) {
            if (i==0 && j==0) {
                dp[i][j] = 0;               // edge case
            } else if (i == 0) {
                dp[i][j] = dp[i][j-1]+1;    // insert/delete
            } else if (j == 0) {
                dp[i][j] = dp[i-1][j]+1;    // insert/delete
            } else if (word1[i-1] == word2[j-1]) {
                dp[i][j] = dp[i-1][j-1];    // as same as (w1[i-2], w2[j-2])
            } else {
                dp[i][j] = min({dp[i-1][j-1], dp[i-1][j], dp[i][j-1]})+1; // min from replace, insert
            }
        }
        }
        return dp[A][B];
    }
};
// @lc code=end

